### PR TITLE
Log 2118: Checking for PVC Existence Before Create

### PR DESCRIPTION
**Pull request by Red-GV**
This PR causes the operator to check for the existence of the PVC before attempting to create it. ResourceQuota errors are returned as a 403, so we cannot fire the create method first before checking without creating a PVC during every reconcile period.

/cc @sasagarw 
/assign @jcantrill 

### Links
- JIRA: https://issues.redhat.com/browse/LOG-2118

### CODE REVIEW

File: internal/elasticsearch/common.go [hunk 1]
```diff
@@ -637,14 +637,49 @@ func newVolumeSource(clusterName, nodeName, namespace string, node api.Elasticse
 		StorageClassName: specVol.StorageClassName,
 	}
 
-	err := persistentvolume.CreateOrUpdatePVC(context.TODO(), client, pvc, persistentvolume.LabelsEqual, persistentvolume.MutateLabelsOnly)
+	// PersistentVolumeClaims can run into ResourceQuota error. Since the create
```
Comment (jcantrill) [inline internal/elasticsearch/common.go:640]: Why are we hitting the API server in what amounts to a contructor (e.g. newXXXXX).  The original method should have never been their either

File: internal/elasticsearch/common.go [hunk 2]
```diff
@@ -637,14 +637,49 @@ func newVolumeSource(clusterName, nodeName, namespace string, node api.Elasticse
...
+		err = persistentvolume.CreateOrUpdatePVC(context.TODO(), client, pvc, persistentvolume.LabelsEqual, persistentvolume.MutateLabelsOnly)
+		if err != nil {
+			log.Error(err, "Unable to create PersistentVolumeClaim")
+		}
 	}
 
 	return volSource
 }
 
+func doesPersistentVolumeClaimExist(claimName, namespace string, c client.Client) (bool, error) {
```
Comment (jcantrill) [inline internal/elasticsearch/common.go:665]: Why is it important to return a bool and the error?  Can we not return only a bool and log the error if one occurs?

File: internal/elasticsearch/common.go [hunk 3]
```diff
@@ -637,14 +637,49 @@ func newVolumeSource(clusterName, nodeName, namespace string, node api.Elasticse
...
-	err := persistentvolume.CreateOrUpdatePVC(context.TODO(), client, pvc, persistentvolume.LabelsEqual, persistentvolume.MutateLabelsOnly)
+	// PersistentVolumeClaims can run into ResourceQuota error. Since the create
+	// method will return a 403 when it runs into this issue, the existing pvc
+	// should be found and updated rather than attempt to create it and catch the
+	// already exists error.
+	status, err := doesPersistentVolumeClaimExist(claimName, namespace, client)
+
 	if err != nil {
-		log.Error(err, "Unable to create PersistentVolumeClaim")
+		log.Error(err, "Unable to get PersistentVolumeClaim")
```
Comment (jcantrill) [inline internal/elasticsearch/common.go:647]: Complimenting the other comment.  Why  can this error not be pushed into the other function?

File: internal/elasticsearch/common.go [hunk 4]
```diff
@@ -637,14 +637,49 @@ func newVolumeSource(clusterName, nodeName, namespace string, node api.Elasticse
...
+	}
+
+	if status {
```
Comment (jcantrill) [inline internal/elasticsearch/common.go:650]: This in general seems like a code smell.... why can "CreateOrUpdatePVC" not do as advertised by its name?  If it doesn't createorupdate... then lets make it do that.

File: internal/manifests/persistentvolume/persistentvolumeclaim.go [hunk 5]
```diff
@@ -74,6 +59,26 @@ func CreateOrUpdatePVC(ctx context.Context, c client.Client, pvc *corev1.Persist
 	return nil
 }
 
+// CreateOrUpdatePVC attempts first to create the given persistentvolumeclaim. If the
+// persistentvolumeclaim already exists and the provided comparison func detects any changes
+// an update is attempted. Updates are retried with backoff (See retry.DefaultRetry).
+// Returns on failure an non-nil error.
+func CreateOrUpdatePVC(ctx context.Context, c client.Client, pvc *corev1.PersistentVolumeClaim, equal EqualityPVCFunc, mutate MutatePVCFunc) error {
+	err := c.Create(ctx, pvc)
```
Comment (jcantrill) [inline internal/manifests/persistentvolume/persistentvolumeclaim.go:67]: I just changed this a function in the CLO and this pattern is pervasive in both our operators but it is backwords.  The majority of the time we will get/compare/update with less frequently creating.  This should be the default pattern:
```
get
if not found
    create
else
   if compare different
      update
```

Comment (Red-GV) [reply internal/manifests/persistentvolume/persistentvolumeclaim.go:67]: I see. I think after a refactor on our end, a number of CreateOrUpdate functions got turned into

```
create
get
if compare different
   update
```

Changing it to that in the original function, should just clean up your previous options so I'll just go for the default pattern and then write a card to evaluate how many of our functions need to have this change in the pattern done as well.

**Review by sasagarw — APPROVED** [review]
/lgtm

**Pull request merged**